File: internal/workspace/workspace.go
package workspace

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type Workspace struct {
	dir string
}

func Open(dir string) (*Workspace, error) {
	w := &Workspace{dir: dir}
	if _, err := w.git(nil, "rev-parse", "--git-dir"); err != nil {
		if _, err := w.git(nil, "init"); err != nil {
			return nil, fmt.Errorf("initializing git repository: %w", err)
		}
	}
	if err := w.excludeFromSnapshots("/.nina/"); err != nil {
		return nil, err
	}
	return w, nil
}

func (w *Workspace) excludeFromSnapshots(pattern string) error {
	gitDir, err := w.git(nil, "rev-parse", "--git-dir")
	if err != nil {
		return err
	}
	if !filepath.IsAbs(gitDir) {
		gitDir = filepath.Join(w.dir, gitDir)
	}
	path := filepath.Join(gitDir, "info", "exclude")
	existing, err := os.ReadFile(path)
	if err != nil && !os.IsNotExist(err) {
		return err
	}
	for line := range strings.SplitSeq(string(existing), "\n") {
		if strings.TrimSpace(line) == pattern {
			return nil
		}
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	content := string(existing)
	if content != "" && !strings.HasSuffix(content, "\n") {
		content += "\n"
	}
	return os.WriteFile(path, []byte(content+pattern+"\n"), 0o644)
}

func SnapshotRef(session string, step int) string {
	return fmt.Sprintf("refs/nina/%s/%d", session, step)
}

func (w *Workspace) Snapshot(ref string) (string, error) {
	indexDir, err := os.MkdirTemp("", "nina-index-*")
	if err != nil {
		return "", err
	}
	defer os.RemoveAll(indexDir)
	indexEnv := []string{"GIT_INDEX_FILE=" + filepath.Join(indexDir, "index")}

	if _, err := w.git(indexEnv, "add", "-A", "."); err != nil {
		return "", fmt.Errorf("staging snapshot: %w", err)
	}
	tree, err := w.git(indexEnv, "write-tree")
	if err != nil {
		return "", fmt.Errorf("writing snapshot tree: %w", err)
	}
	commit, err := w.git(nil, "commit-tree", tree, "-m", "nina snapshot")
	if err != nil {
		return "", fmt.Errorf("writing snapshot commit: %w", err)
	}
	if _, err := w.git(nil, "update-ref", ref, commit); err != nil {
		return "", fmt.Errorf("updating snapshot ref: %w", err)
	}
	return commit, nil
}

func (w *Workspace) Diff(refA, refB string) (string, error) {
	return w.git(nil, "diff", refA, refB)
}

func (w *Workspace) git(extraEnv []string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = w.dir
	cmd.Env = append(os.Environ(),
		"GIT_AUTHOR_NAME=nina",
		"GIT_AUTHOR_EMAIL=nina@localhost",
		"GIT_COMMITTER_NAME=nina",
		"GIT_COMMITTER_EMAIL=nina@localhost",
	)
	cmd.Env = append(cmd.Env, extraEnv...)
	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("git %s: %w: %s", strings.Join(args, " "), err, strings.TrimSpace(stderr.String()))
	}
	return strings.TrimSpace(stdout.String()), nil
}

File: python/src/nina/workspace.py
from __future__ import annotations

import os
import subprocess
import tempfile
from pathlib import Path


class WorkspaceError(RuntimeError):
    pass


class Workspace:
    def __init__(self, dir: str) -> None:
        self.dir = dir

    def _git(self, args: list[str], extra_env: dict[str, str] | None = None) -> str:
        env = os.environ.copy()
        env.update(
            {
                "GIT_AUTHOR_NAME": "nina",
                "GIT_AUTHOR_EMAIL": "nina@localhost",
                "GIT_COMMITTER_NAME": "nina",
                "GIT_COMMITTER_EMAIL": "nina@localhost",
            }
        )
        if extra_env:
            env.update(extra_env)
        proc = subprocess.run(
            ["git", *args],
            cwd=self.dir,
            env=env,
            capture_output=True,
            text=True,
        )
        if proc.returncode != 0:
            raise WorkspaceError(f"git {' '.join(args)}: {proc.stderr.strip()}")
        return proc.stdout.strip()

    def _exclude_from_snapshots(self, pattern: str) -> None:
        git_dir = Path(self._git(["rev-parse", "--git-dir"]))
        if not git_dir.is_absolute():
            git_dir = Path(self.dir) / git_dir
        path = git_dir / "info" / "exclude"
        existing = path.read_text() if path.exists() else ""
        if any(line.strip() == pattern for line in existing.split("\n")):
            return
        path.parent.mkdir(parents=True, exist_ok=True)
        content = existing
        if content and not content.endswith("\n"):
            content += "\n"
        path.write_text(content + pattern + "\n")

    def snapshot(self, ref: str) -> str:
        with tempfile.TemporaryDirectory(prefix="nina-index-") as index_dir:
            index_env = {"GIT_INDEX_FILE": str(Path(index_dir) / "index")}
            self._git(["add", "-A", "."], extra_env=index_env)
            tree = self._git(["write-tree"], extra_env=index_env)
        commit = self._git(["commit-tree", tree, "-m", "nina snapshot"])
        self._git(["update-ref", ref, commit])
        return commit

    def diff(self, ref_a: str, ref_b: str) -> str:
        return self._git(["diff", ref_a, ref_b])


def open_workspace(dir: str) -> Workspace:
    w = Workspace(dir)
    try:
        w._git(["rev-parse", "--git-dir"])
    except WorkspaceError:
        try:
            w._git(["init"])
        except WorkspaceError as err:
            raise WorkspaceError(f"initializing git repository: {err}") from err
    w._exclude_from_snapshots("/.nina/")
    return w


def snapshot_ref(session: str, step: int) -> str:
    return f"refs/nina/{session}/{step}"
